File: finalyearproject/repos/dkd-platform-libs/sdk/typescript/src/idempotency.ts
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// Idempotency-Key enforcement (PL-03) for unsafe/money/custody writes (EF-API-6 / EF §21.1):
//   - a write MISSING an Idempotency-Key            → HTTP 400 (Bad Request)
//   - a replay: SAME key + SAME request payload     → the ORIGINAL stored response is returned
//   - a conflict: SAME key + DIFFERENT payload       → HTTP 409 (Conflict)
// Backed by a PLUGGABLE store (the PL-02 inbox / a dedicated idempotency table) — the DB is NOT
// hard-wired here. One of five language implementations sharing the SAME contract.

import { createHash } from "node:crypto";

/** HTTP methods that MUST carry an Idempotency-Key (unsafe/money/custody writes). */
const UNSAFE_METHODS: ReadonlySet<string> = new Set(["POST", "PUT", "PATCH", "DELETE"]);

/** True when the method mutates state and therefore requires an Idempotency-Key. */
export function isUnsafeMethod(method: string): boolean {
  return UNSAFE_METHODS.has(method.toUpperCase());
}

/** SHA-256 (lowercase hex) of a request payload — the fingerprint compared on replay. */
export function fingerprint(payload: string | Uint8Array): string {
  return createHash("sha256").update(payload).digest("hex");
}

/** The record persisted for a completed idempotent request (the response to replay). */
export interface IdempotencyRecord {
  key: string;
  /** {@link fingerprint} of the original request payload. */
  payloadHash: string;
  statusCode: number;
  /** Serialized original response body, replayed verbatim. */
  responseBody: string;
  createdAtMs: number;
}

/**
 * Pluggable persistence for idempotency records. Back it with the PL-02 inbox table or a
 * dedicated `idempotency_keys` table — the helper stays storage-agnostic.
 */
export interface IdempotencyStore {
  /** Fetch a previously stored record for `key`, or null when unseen. */
  find(key: string): Promise<IdempotencyRecord | null>;
  /** Persist a record. Returns false when `key` already existed (lost a concurrent race). */
  save(record: IdempotencyRecord): Promise<boolean>;
}

/** Outcome of {@link decide} — a discriminated union the caller acts on. */
export type IdempotencyOutcome =
  /** No Idempotency-Key on an unsafe write → respond 400. */
  | { kind: "missing-key"; status: 400 }
  /** Same key + same payload → replay the stored response verbatim. */
  | { kind: "replay"; status: number; record: IdempotencyRecord }
  /** Same key + different payload → respond 409. */
  | { kind: "conflict"; status: 409 }
  /** Unseen key → proceed with the write, then persist via {@link IdempotencyStore.save}. */
  | { kind: "proceed"; key: string; payloadHash: string };

export interface IdempotencyRequest {
  method: string;
  /** The `Idempotency-Key` header value (undefined when absent). */
  key?: string;
  /** The raw request payload (body) used to compute the fingerprint. */
  payload: string | Uint8Array;
}

/**
 * Enforce the idempotency contract for one request. Safe methods (GET/HEAD/...) always proceed
 * without a key. This is transport-agnostic: HTTP frameworks map the outcome to a response;
 * the three-branch enforcement logic lives here and is unit-tested against a fake store.
 */
export async function decide(
  store: IdempotencyStore,
  req: IdempotencyRequest,
): Promise<IdempotencyOutcome> {
  if (!isUnsafeMethod(req.method)) {
    return { kind: "proceed", key: req.key ?? "", payloadHash: "" };
  }
  if (!req.key) {
    return { kind: "missing-key", status: 400 };
  }
  const payloadHash = fingerprint(req.payload);
  const existing = await store.find(req.key);
  if (existing) {
    if (existing.payloadHash === payloadHash) {
      return { kind: "replay", status: existing.statusCode, record: existing };
    }
    return { kind: "conflict", status: 409 };
  }
  return { kind: "proceed", key: req.key, payloadHash };
}

/**
 * Persist the response produced for a `proceed` outcome so a later replay returns it verbatim.
 * Call this after the handler has produced its response. Returns the stored record.
 */
export async function remember(
  store: IdempotencyStore,
  outcome: Extract<IdempotencyOutcome, { kind: "proceed" }>,
  statusCode: number,
  responseBody: string,
  nowMs: number = Date.now(),
): Promise<IdempotencyRecord> {
  const record: IdempotencyRecord = {
    key: outcome.key,
    payloadHash: outcome.payloadHash,
    statusCode,
    responseBody,
    createdAtMs: nowMs,
  };
  await store.save(record);
  return record;
}

File: finalyearproject/repos/dkd-platform-libs/sdk/go/idempotency.go
// Hand-authored primitive (PL-03). NOT dkdgen output: the Idempotency-Key HTTP
// middleware enforcing canon EF-API-6 on unsafe/money/custody writes —
//   - MISSING key on an unsafe write            -> 400 (malformed)
//   - SAME key + SAME payload (replay)           -> the stored original response
//   - SAME key + DIFFERENT payload               -> 409 (state conflict)
// It is backed by a pluggable IdempotencyStore (the PL-02 inbox / a dedicated
// idem table / an in-memory fake) — no DB is hard-wired here.

package dkdplatform

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"io"
	"net/http"
	"sync"
)

// IdempotencyKeyHeader is the canonical request header (EF-API-6).
const IdempotencyKeyHeader = "Idempotency-Key"

// IdempotencyRecord is the stored original response for a given key, plus the
// hash of the request payload that produced it (to detect a same-key/different-
// payload replay).
type IdempotencyRecord struct {
	PayloadHash string
	Status      int
	Body        []byte
	Headers     map[string]string
}

// IdempotencyStore is the pluggable persistence seam. Lookup reports (record,
// found, error); Save records the first response under a key. A DB-backed store
// implements this over an idem table in the same tx discipline as the PL-02
// inbox; MemIdempotencyStore is the in-memory implementation used in tests.
type IdempotencyStore interface {
	Lookup(ctx context.Context, key string) (IdempotencyRecord, bool, error)
	Save(ctx context.Context, key string, rec IdempotencyRecord) error
}

// MemIdempotencyStore is a goroutine-safe in-memory IdempotencyStore. It is the
// reference/fake implementation — production services supply a durable store.
type MemIdempotencyStore struct {
	mu   sync.Mutex
	data map[string]IdempotencyRecord
}

// NewMemIdempotencyStore constructs an empty in-memory store.
func NewMemIdempotencyStore() *MemIdempotencyStore {
	return &MemIdempotencyStore{data: map[string]IdempotencyRecord{}}
}

func (s *MemIdempotencyStore) Lookup(_ context.Context, key string) (IdempotencyRecord, bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	rec, ok := s.data[key]
	return rec, ok, nil
}

func (s *MemIdempotencyStore) Save(_ context.Context, key string, rec IdempotencyRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, exists := s.data[key]; exists {
		return nil // first-writer-wins; a concurrent duplicate is a no-op
	}
	s.data[key] = rec
	return nil
}

// isUnsafeMethod reports whether a method mutates state and therefore requires
// an Idempotency-Key (EF-API-6). GET/HEAD/OPTIONS/TRACE are safe.
func isUnsafeMethod(m string) bool {
	switch m {
	case http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete:
		return true
	default:
		return false
	}
}

// IdempotencyMiddleware wraps a handler with EF-API-6 enforcement backed by the
// given store. Safe methods pass straight through. For an unsafe method it
// requires the Idempotency-Key header, replays the stored response on a
// same-payload retry, and returns 409 on a same-key/different-payload clash.
func IdempotencyMiddleware(store IdempotencyStore) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if !isUnsafeMethod(r.Method) {
				next.ServeHTTP(w, r)
				return
			}
			key := r.Header.Get(IdempotencyKeyHeader)
			if key == "" {
				WriteProblem(w, NewMalformedError(
					"dokandar.platform.idempotency.missing_key",
					"Idempotency-Key header is required on this request"), 0)
				return
			}

			body, err := io.ReadAll(r.Body)
			if err != nil {
				WriteProblem(w, NewMalformedError(
					"dokandar.platform.idempotency.unreadable_body",
					"request body could not be read"), 0)
				return
			}
			_ = r.Body.Close()
			r.Body = io.NopCloser(bytes.NewReader(body))
			hash := hashPayload(body)

			rec, found, err := store.Lookup(r.Context(), key)
			if err != nil {
				WriteProblem(w, NewUnavailableError(
					"dokandar.platform.idempotency.store_unavailable",
					"idempotency store is unavailable"), 0)
				return
			}
			if found {
				if rec.PayloadHash == hash {
					replay(w, rec)
					return
				}
				WriteProblem(w, NewConflictError(
					"dokandar.platform.idempotency.payload_mismatch",
					"Idempotency-Key reused with a different request payload"), 0)
				return
			}

			rec = IdempotencyRecord{PayloadHash: hash}
			rw := &recordingWriter{ResponseWriter: w, status: http.StatusOK}
			next.ServeHTTP(rw, r)
			rec.Status = rw.status
			rec.Body = rw.body.Bytes()
			rec.Headers = flattenHeaders(rw.Header())
			_ = store.Save(r.Context(), key, rec)
		})
	}
}

func hashPayload(b []byte) string {
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:])
}

func replay(w http.ResponseWriter, rec IdempotencyRecord) {
	for k, v := range rec.Headers {
		w.Header().Set(k, v)
	}
	w.Header().Set("Idempotency-Replayed", "true")
	if rec.Status == 0 {
		rec.Status = http.StatusOK
	}
	w.WriteHeader(rec.Status)
	_, _ = w.Write(rec.Body)
}

func flattenHeaders(h http.Header) map[string]string {
	out := make(map[string]string, len(h))
	for k := range h {
		out[k] = h.Get(k)
	}
	return out
}

// recordingWriter captures the handler's status + body so the original response
// can be stored and later replayed.
type recordingWriter struct {
	http.ResponseWriter
	status      int
	body        bytes.Buffer
	wroteHeader bool
}

func (rw *recordingWriter) WriteHeader(code int) {
	if rw.wroteHeader {
		return
	}
	rw.status = code
	rw.wroteHeader = true
	rw.ResponseWriter.WriteHeader(code)
}

func (rw *recordingWriter) Write(b []byte) (int, error) {
	if !rw.wroteHeader {
		rw.WriteHeader(http.StatusOK)
	}
	rw.body.Write(b)
	return rw.ResponseWriter.Write(b)
}
